src/site/topic15.rst

```diff
@@ -3,6 +3,69 @@ Topic #15 --- ArrayLists
 ************************
 
 
+ArraySortedBag
+--------------
+
+* In addition to the functionality of the ``Bag``, we want to add a few more methods for our ``SortedBag`` implementation
+    * ``void add(T element)`` --- add to the proper spot in the sorted bag to preserve the sorted order
+    * ``T removeFirst()``
+    * ``T removeLast()``
+    * ``T first()``
+    * ``T last()``
+
+.. warning::
+
+    Like the indexed bag, some methods are skipped. See the
+    :download:`ArraySortedBag <../main/java/ArraySortedBag.java>` implementation to view the full implementation.
+
+.. code-block:: Java
+    :linenos:
+    :emphasize-lines: 4
+
+    import java.util.Iterator;
+    import java.util.NoSuchElementException;
+
+    public class ArraySortedBag<T extends Comparable<? super T>> implements SortedBag<T> {
+
+        private static final int DEFAULT_CAPACITY = 100;
+        private static final int NOT_FOUND = -1;
+        protected T[] bag;
+        protected int rear;
+
+        public ArraySortedBag() {
+            this(DEFAULT_CAPACITY);
+        }
+
+        @SuppressWarnings("unchecked")
+        public ArraySortedBag(int initialCapacity) {
+            bag = (T[]) new Comparable[initialCapacity];
+            rear = 0;
+        }
+
+* This looks nearly the same as the ``ArrayIndexedBag`` implementation, but we see one major difference
+
+    ``<T extends Comparable<? super T>>``
+
+* There is a little bit to unpack here
+
+* First, when something extends `Comparable <https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Comparable.html>`_, it means that the type has some defined ordering
+    * The method ``compareTo`` is implemented for the type
+
+* If we call ``x.compareTo(y)``
+    * Return a negative integer if ``y < x``
+    * Return zero if ``y == x``
+    * Return a positive integer if ``y > x``
+
+* By saying ``T extends Comparable``, this means that the type ``T`` we want in our ``ArraySortedBag`` has some defined ordering
+    * Which is needed, if we want to sort things
```

Comment:
Wording is slightly off here. A `Comparable<T>` means you can compare `this` to some type `T` -- you can do `class Thing extends Comparable<NotThing>` to say _things can be compared to not things_, but not the other way around. `T extends Comparable<T>` means that `T` can be compared to `T` -- ie, we want a type that I can compare to the same type, defining a total ordering. Then the `? super T` would say that `T` can be compared to some supertype of `T` (or `T` itself), which you explained fairly well here, allowing for inheriting the `compareTo` method.